### PR TITLE
XIVY-13646 Add DataTable-Component

### PR DESCRIPTION
**Pull request by ivy-edp**
I implemented the first version of the DataTable component in the form editor. It can:
- Add new columns via quick actions of DataTable
- Move columns within DataTables
- Delete columns via delete or quick action of Column
- Show a button to define a list of objects when the DataTable value is empty (a warning appears if the value is empty)
- Generate columns based on the DataTable-value if something is defined (only show Button if DataTable-value is defined)
- Define column value via browser if a list of objects is set within the DataTable

Open To-Dos:
- Generation doesn’t work for lists of strings or primitive data types
- Column movement across different DataTables should be disabled
- Tests (UT/IT)

firstOfDataTable

### CODE REVIEW

File: packages/editor/src/editor/canvas/ComponentBlock.tsx [hunk 1]
```diff
@@ -78,7 +88,16 @@ const Draggable = ({ config, data }: DraggableProps) => {
...
+      {data.type === 'DataTableColumn' ? (
+        <Quickbar deleteAction={deleteElement} />
+      ) : (
+        <Quickbar
+          deleteAction={deleteElement}
+          duplicateAction={duplicateElement}
+          createAction={createElement}
+          createColumnAction={data.type === 'DataTable' ? createColumn : undefined}
+        />
+      )}
```
Comment (ivy-lli) [inline packages/editor/src/editor/canvas/ComponentBlock.tsx:100]: seems also wrong to me that the componentblock needs to check what actions a type has... This should be defined in the config of the component itself...

File: packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  outlineInfo: component => component.header,
+  fields: {
+    header: { subsection: 'General', label: 'Header', type: 'text' },
+    value: { subsection: 'General', label: 'Value', type: 'textBrowser', options: { onlyAttributes: 'COLUMN' } },
+    ...baseComponentFields
+  }
+};
+
+const UiBlock = ({ header, value }: UiComponentProps<DataTableColumnProps>) => (
+  <>
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.tsx:34]: fragmet not needed.

File: packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.svg
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.svg:1]: no longer needed I think, as we do not render this on the palette.

File: packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.css
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.css:1]: remove empty file...

File: packages/editor/src/components/blocks/datatable/DataTable.tsx [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    components: { subsection: 'General', type: 'hidden' },
+    label: { subsection: 'General', label: 'Label', type: 'text' },
+    value: { subsection: 'General', label: 'Value', type: 'textBrowser', options: { onlyTypesOf: 'List<' } },
+    ...baseComponentFields
+  }
+};
+
+const UiBlock = ({ id, label, components, value }: UiComponentProps<DataTableProps>) => {
+  return (
+    <>
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatable/DataTable.tsx:39]: fragmet not needed.
use short return syntax `=> ()`

File: packages/editor/src/components/blocks/datatable/DataTable.tsx [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+          </span>
+        </div>
+        {components.length > 0 && (
+          <div className='block-table__columns'>
+            {components.map((column, index) => {
+              return <ComponentBlock key={column.id} component={column} preId={components[index - 1]?.id} />;
+            })}
+          </div>
+        )}
+        {components.length === 0 && <DataTableColumnDialog id={id} initValue={value} />}
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatable/DataTable.tsx:54]: I wouldn't do this...
Just render a warning which is fine, and if someone wants to define this then go to the property view. -> Keep it simple and stupid

File: packages/editor/src/components/blocks/datatable/DataTable.tsx [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+        <div className='block-table__label'>
+          <span>{label}</span>
+          <span style={{ color: 'var(--N600)' }}>
+            {value.length > 0 ? value : <Message variant='warning' message='Value of DataTable is empty' />}
+          </span>
+        </div>
+        {components.length > 0 && (
+          <div className='block-table__columns'>
+            {components.map((column, index) => {
+              return <ComponentBlock key={column.id} component={column} preId={components[index - 1]?.id} />;
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatable/DataTable.tsx:50]: use short return syntax `=> <ComponentBlock...`

File: packages/editor/src/data/data.ts [hunk 6]
```diff
@@ -21,14 +22,17 @@ const findComponent = (
   if (id.startsWith(LAYOUT_DROPZONE_ID_PREFIX)) {
     return findLayoutComponent(data, id.replace(LAYOUT_DROPZONE_ID_PREFIX, ''));
   }
+  if (id.startsWith(TABLE_DROPZONE_ID_PREFIX)) {
+    return findTableComponent(data, id.replace(TABLE_DROPZONE_ID_PREFIX, ''));
+  }
   return findComponentDeep(data, id, parent);
```
Comment (ivy-lli) [inline packages/editor/src/data/data.ts:28]: instead of mix this table component stuff with the normal data behaviour, I just had an idea...
What about create an own DndContext just for the datatable columns. Then you maybe don't have to handle stuff like that something else can be dropped into the table or that a column can be dropped outside...

File: packages/editor/src/editor/canvas/data-table-column/DataTableColumnDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+import { flexRender, type ColumnDef, useReactTable, getCoreRowModel, getFilteredRowModel } from '@tanstack/react-table';
+import { useMeta } from '../../../context/useMeta';
+import { useAppContext } from '../../../context/AppContext';
+import { useEffect, useState } from 'react';
+import type { Variable } from '@axonivy/form-editor-protocol';
+import { findAttributesOfType, findVariablesOfType } from '../../../data/variable-tree-data';
+import { createInitiTableColumns, useData } from '../../../data/data';
+import { componentByName } from '../../../components/components';
+
+export const DataTableColumnDialog = ({ id, initValue }: { id: string; initValue: string }) => {
```
Comment (ivy-lli) [inline packages/editor/src/editor/canvas/data-table-column/DataTableColumnDialog.tsx:27]: If you keep it simple that a user needs to define this value via the property view then this whole thing is not needed...

File: packages/editor/src/components/blocks/datatable/DataTable.tsx (hunk 4 above)
Comment (ivy-lli) [reply packages/editor/src/components/blocks/datatable/DataTable.tsx:54]: Or better just provide a button here if the value is set to create the columns but not a button for the dialog...

File: packages/editor/src/editor/canvas/data-table-column/DataTableColumnDialog.tsx (hunk 7 above)
Comment (ivy-lli) [reply packages/editor/src/editor/canvas/data-table-column/DataTableColumnDialog.tsx:27]: at least the browser not, and the other thing is only a button which creates the columns

File: packages/editor/src/editor/sidebar/fields/browser/useAttributeBrowser.tsx [hunk 8]
```diff
@@ -3,28 +3,47 @@ import { IvyIcons } from '@axonivy/ui-icons';
...
+          const hasMatchingColumn = (component.config as unknown as DataTable).components.some(
+            childComponent => childComponent.id === element?.id
+          );
+          if (hasMatchingColumn) {
+            return component.config as unknown as DataTable;
+          }
+        }
+      }
+      return undefined;
+    };
```
Comment (ivy-lli) [inline packages/editor/src/editor/sidebar/fields/browser/useAttributeBrowser.tsx:33]: why is this function here? has nothing to do with this useEffect right? Only here to find the parent datatable.

File: packages/editor/src/editor/sidebar/fields/browser/useAttributeBrowser.tsx [hunk 9]
```diff
@@ -3,28 +3,47 @@ import { IvyIcons } from '@axonivy/ui-icons';
...
+
+    const parentTableComponent = findParentTableComponent();
```
Comment (ivy-lli) [inline packages/editor/src/editor/sidebar/fields/browser/useAttributeBrowser.tsx:35]: This code is now called for every browser usage, even if it has nothing to do with a table...
This should only be executed if a column filter is active!

**Review by ivy-lli — CHANGES_REQUESTED** [review]

File: packages/editor/src/data/data.ts (hunk 6 above)
Comment (ivy-edp) [reply packages/editor/src/data/data.ts:28]: Thats a good idea 👍 would it be possible to try this in a later PR. Maybe all this stuff will change anyway as i implement in the next PR the updated/improved form.ts

File: packages/editor/src/editor/canvas/drag-data.ts [hunk 10]
```diff
@@ -52,3 +44,16 @@ export const isDropZoneDisabled = (id: string, active: Active | null, preId?: st
   }
   return false;
 };
+
+const disableDropZoneDataTableColumn = (id: string, active: Active | null) => {
```
[the remark this answers is not in the document]
Comment (ivy-lli) [reply packages/editor/src/editor/canvas/drag-data.ts:48]: second: still hard to read... why not (only in my head, not sure if correct):

```ts
const isColumn = id.startsWith('DataTableColumn');
const isColumnTarget = dropZoneId.includes('DataTableColumn');
if (isColumn) {
  return isColumnTarget || dropZoneId.startWith(TABLE_DROPZONE_ID_PREFIX);
} 
return !isColumnTarget;
```

File: packages/editor/src/editor/palette/Palette.tsx [hunk 11]
```diff
@@ -14,7 +14,9 @@ export const Palette = ({ sections, directCreate }: PaletteProps) => {
     <Flex direction='column' className='palette' gap={3}>
       <SearchInput placeholder='Search...' value={searchTerm} onChange={setSearchTerm} />
       {Object.entries(sections).map(([section, sectionItems]) => {
-        const filteredItems = sectionItems.filter(item => item.name.toLowerCase().includes(searchTerm.toLowerCase()));
+        const filteredItems = sectionItems
+          .filter(item => item.name !== 'DataTableColumn')
+          .filter(item => item.name.toLowerCase().includes(searchTerm.toLowerCase()));
```
Comment (ivy-lli) [inline packages/editor/src/editor/palette/Palette.tsx:19]: Not sure if you understand what I mean...
I think this should not be filtered by the palette components. Or at least not by the item name.
What if this item name changes one day (what I think is not too unrealistic)? -> bug

Why not add a new category "hidden"? then the filtering here would make much more sense.

**Review by ivy-lli — CHANGES_REQUESTED** [review]
Much better, but still potential for improvements!

File: packages/editor/src/editor/canvas/ComponentBlock.tsx (hunk 1 above)
Comment (ivy-lli) [reply packages/editor/src/editor/canvas/ComponentBlock.tsx:100]: -> new subtask

File: packages/editor/src/data/data.ts (hunk 6 above)
Comment (ivy-lli) [reply packages/editor/src/data/data.ts:28]: -> new subtask

File: packages/editor/src/components/blocks/datatable/DataTable.tsx [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        {components.map((column, index) => (
+          <ComponentBlock key={column.id} component={column} preId={components[index - 1]?.id} />
+        ))}
+      </div>
+    )}
+    {components.length === 0 && <DataTableColumnDialog id={id} initValue={value} />}
+  </div>
+);
+
+const DataTableColumnDialog = ({ id, initValue }: { id: string; initValue: string }) => {
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatable/DataTable.tsx:60]: And now you can also rename this component to something like `EmptyDataTableColumn`

File: packages/editor/src/components/blocks/datatable/DataTable.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+  const { context, setData } = useAppContext();
+  const [tree, setTree] = useState<Array<BrowserNode<Variable>>>([]);
+  const dataClass = useMeta('meta/data/attributes', context, { types: {}, variables: [] }).data;
+  useEffect(() => setTree(findAttributesOfType(dataClass, initValue)), [dataClass, initValue]);
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatable/DataTable.tsx:64]: I don't think that you really need the tree state and the useEffect.
You only use them when you click on the create button, so this findAttributesType logic can also be directly in the createColumns function right?

File: packages/editor/src/components/blocks/datatable/DataTable.tsx [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+    });
+  };
+
+  return initValue.length > 0 ? (
+    <Button icon={IvyIcons.DatabaseLink} variant='outline' onClick={createColumns}>
+      Create columns from value
+    </Button>
+  ) : (
+    <Message variant='warning' message='Value of DataTable is empty. Define value within Properties of DataTable' />
+  );
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatable/DataTable.tsx:91]: instead of this, I would fast return, e.g. like:
```tsx
if (initValue.length === 0) {
  return <Message ... />
}

const createColumns = () => {
...
}

return <Button ... />

```

File: packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.tsx [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+  defaultProps: defaultDataTableColumnProps,
+  render: props => <UiBlock {...props} />,
+  create: ({ label, value }) => ({ ...defaultDataTableColumnProps, header: label, value }),
+  outlineInfo: component => component.header,
+  fields: {
+    header: { subsection: 'General', label: 'Header', type: 'text' },
+    value: { subsection: 'General', label: 'Value', type: 'textBrowser', options: { onlyAttributes: 'COLUMN' } },
+    ...baseComponentFields
+  },
+  hidden: true
```
Comment (ivy-lli) [inline packages/editor/src/components/blocks/datatablecolumn/DataTableColumn.tsx:29]: Better, but still not 100% was I had in mind.
I meant that you instead of add this block to the category 'Elements', add it to a category 'Hidden'.

File: packages/editor/src/components/components.test.ts [hunk 16]
```diff
@@ -11,15 +11,15 @@ test('componentByName', () => {
...
   const result = componentsByCategory('Elements');
   expect(result.General).toEqual(undefined);
-  expect(result.Input).toHaveLength(4);
+  expect(result.Input).toHaveLength(5);
   expect(result.Selection).toHaveLength(3);
   expect(result.Text).toHaveLength(1);
 });
 
 test('allComponentsByCategory', () => {
   const result = allComponentsByCategory();
```
Comment (ivy-lli) [inline packages/editor/src/components/components.test.ts:20]: If you use the hidden category (other comment) you could add a simple check here that the hidden category will not appear in this object.

File: packages/editor/src/editor/canvas/drag-data.ts [hunk 17]
```diff
@@ -35,3 +44,13 @@ export const isDropZoneDisabled = (id: string, active: Active | null, preId?: st
...
+
+const disableDropZoneDataTableColumn = (id: string, dropZoneId: string | undefined) => {
+  const isColumn = id.startsWith('DataTableColumn');
+  const isColumnTarget = dropZoneId ? dropZoneId.includes('DataTableColumn') : false;
+
+  if (isColumn) {
+    return !isColumnTarget;
+  }
+  return isColumnTarget;
+};
```
Comment (ivy-lli) [inline packages/editor/src/editor/canvas/drag-data.ts:56]: 👍 
a lot easier or what do you think?

File: packages/editor/src/editor/palette/Palette.tsx
Comment (ivy-lli) [inline packages/editor/src/editor/palette/Palette.tsx:1]: you can revert this change

**Review by ivy-lli — COMMENTED** [review]
Again a lot better, however some suggestions:

**Review by ivy-lli — APPROVED** [review]
Thanks for the improvements.
Sorry to be so accurate 🙈
